libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFNetwork.h

```diff
@@ -32,31 +32,46 @@
 
 NS_SWIFT_NAME(NetworkManaging)
 @protocol SFNetworkSessionManaging
-- (nonnull NSURLSession *)ephemeralSession:(nonnull NSURLSessionConfiguration *)sessionConfig;
-- (nonnull NSURLSession *)backgroundSession:(nonnull NSURLSessionConfiguration *)sessionConfig;
+- (nonnull NSURLSession *)ephemeralSession:(nonnull NSURLSessionConfiguration *)sessionConfig SFSDK_DEPRECATED(7.3, 8.0, "Use sessionWithConfigurationIdentifier:sessionConfiguration:useSharedSession instead");
+- (nonnull NSURLSession *)backgroundSession:(nonnull NSURLSessionConfiguration *)sessionConfig SFSDK_DEPRECATED(7.3, 8.0, "Use sessionWithConfigurationIdentifier:sessionConfiguration:useSharedSession instead");
+- (nonnull NSURLSession *)sessionWithConfigurationIdentifier:(nonnull NSString *)identifier sessionConfiguration:(nullable NSURLSessionConfiguration *)configuration useSharedSession:(BOOL)useSharedSession;
```

Comment:
Instead of having special identifiers that imply certain config, could we keep the old factory methods, and maybe add the useSharedSession:(BOOL)useSharedSession to them?
They would all feed into the new method `(nonnull NSURLSession *)sessionWithConfigurationIdentifier:(nonnull NSString *)identifier sessionConfiguration:(nullable NSURLSessionConfiguration *)configuration useSharedSession:(BOOL)useSharedSession` but at least callers would not need to know about `kSFNetworkEphemeralSessionIdentifier` and `kSFNetworkBackgroundSessionIdentifier`.

Reply:
For that route, should the identifiers still be external for use with `removeSharedSessionForConfigurationIdentifier`? Or should there be convenience wrappers for that too?

Reply:
Hmm, I guess.
By the way, what happens if useSharedSession is false, and you pass an identifier used before?

Reply:
For ephemeral sessions it would just spin up new ones and there wouldn't be a conflict because the identifier wouldn't be used beyond determining the config type... For background sessions there'd be a problem with `[NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:identifier]`, I'm thinking that for the false case we could append the hash like it already is in the existing background init method (thanks for catching!)

Reply:
I agree with Wolf's suggestion. If possible, could we avoid callers having to pass in an identifier?
Some thoughts/suggestions:
- have an extra `useSharedSession` argument that a caller could use to explicitly request a private session by passing in `NO`. If `NO` is passed in, we randomly generate a hash to append to the name (already there), and store it in a `Dictionary` in memory.
- keep the existing API signature (without deprecating), but the implementation would simply call the overloaded method, passing in `useSharedSession` as `YES`.
- add an API to return all available identifier names along with the type of session they are, i.e. `background` or `ephemeral`.
- add an API to remove a session from memory that takes an identifier as an argument.
- add an API to remove all sessions in memory.

Reply:
I think the api is getting a bit complicated to understand. What do you guys think about having  [SFNetwork defaultNetwork] && [SFNetwork defaultBackgroundNetwork] instead for shared mode for salesforce platform? The RestApi will use the above 2 methods for all cases except for external hosts.   Our Auth can create and dispose its own new sessions. Everything else can be custom.

Reply:
@trooper2013 👍 @wmathurin and I were talking about getting rid of the useSharedSession flag yesterday too